zebrad/tests/acceptance.rs

```diff
@@ -1473,3 +1487,413 @@ where
+    network: Network,
+    partial_sync_path: &Path,
+) -> Result<TempDir> {
+    let fully_synced_path = copy_state_directory(&partial_sync_path).await?;
+
+    sync_until(
+        block::Height::MAX,
+        network,
+        SYNC_FINISHED_REGEX,
+        FINISH_PARTIAL_SYNC_TIMEOUT,
+        fully_synced_path,
+        MempoolBehavior::ShouldAutomaticallyActivate,
+        true,
+        false,
+    )
+}
+
+/// Loads transactions from a block that's after the specified `height`.
+///
+/// Starts at the block after the block at the specified `height`, and stops when it finds a block
+/// from where it can load at least one non-coinbase transaction.
+///
+/// # Panics
+///
+/// If the specified `state_path` contains a chain state that's not synchronized to a tip that's
+/// after `height`.
+async fn load_transactions_from_block_after(
+    height: block::Height,
+    network: Network,
+    state_path: &Path,
+) -> Result<Vec<Arc<Transaction>>> {
+    let (_read_write_state_service, mut state, latest_chain_tip, _chain_tip_change) =
+        start_state_service(network, state_path.join("state")).await?;
+
+    let tip_height = latest_chain_tip
+        .best_tip_height()
+        .ok_or_else(|| eyre!("State directory doesn't have a chain tip block"))?;
+
+    assert!(
+        tip_height > height,
+        "Chain not synchronized to a block after the specified height"
+    );
+
+    let mut target_height = height.0;
+    let mut transactions = Vec::new();
+
+    while transactions.is_empty() {
+        transactions =
+            load_transactions_from_block(block::Height(target_height), &mut state).await?;
+
+        transactions.retain(|transaction| !transaction.has_any_coinbase_inputs());
+
+        target_height += 1;
+    }
+
+    Ok(transactions)
+}
+
+/// Performs a request to the provided read-only `state` service to fetch all transactions from a
+/// block at the specified `height`.
+async fn load_transactions_from_block<ReadStateService>(
+    height: block::Height,
+    state: &mut ReadStateService,
+) -> Result<Vec<Arc<Transaction>>>
+where
+    ReadStateService: Service<
+        zebra_state::ReadRequest,
+        Response = zebra_state::ReadResponse,
+        Error = zebra_state::BoxError,
+    >,
+{
+    let request = zebra_state::ReadRequest::Block(HashOrHeight::Height(height));
+
+    let response = state
+        .ready()
+        .and_then(|ready_service| ready_service.call(request))
+        .map_err(|error| eyre!(error))
+        .await?;
+
+    let block = match response {
+        zebra_state::ReadResponse::Block(Some(block)) => block,
+        zebra_state::ReadResponse::Block(None) => {
+            panic!("Missing block at {height:?} from state")
+        }
+        _ => unreachable!("Incorrect response from state service: {response:?}"),
+    };
+
+    Ok(block.transactions.to_vec())
+}
+
+/// Starts a state service using the provided `cache_dir` as the directory with the chain state.
+async fn start_state_service(
```

Comment:
Optional: We could move this utility function to one of the test sub-modules.

Reply:
Done.